contracts/programs/keystone-forwarder/src/utils.rs

```diff
@@ -0,0 +1,43 @@
+use anchor_lang::prelude::Pubkey;
+use anchor_lang::solana_program::hash;
+
+use crate::common::{REPORT_CONTEXT_LEN, SIGNATURE_LEN};
+
+pub fn get_config_id(don_id: u32, config_version: u32) -> u64 {
+    ((don_id as u64) << 32) | (config_version as u64)
+}
+
+// data =  len_signatures (1) | signatures (N*65) | raw_report (M) | report_context (96)
+pub fn extract_raw_report(data: &[u8]) -> &[u8] {
+    let num_signatures = data[0] as usize;
+    let data = &data[1..];
+    let _signatures = &data[..num_signatures * SIGNATURE_LEN];
+    let data = &data[num_signatures * SIGNATURE_LEN..];
+    let _report_context = &data[data.len() - REPORT_CONTEXT_LEN..];
+
+    // raw report
+    &data[..data.len() - REPORT_CONTEXT_LEN]
+}
+
+// version                offset   0, size  1
+// workflow_execution_id  offset   1, size 32
+// timestamp              offset  33, size  4
+// don_id                 offset  37, size  4
+// don_config_version     offset  41, size  4
+// workflow_cid           offset  45, size 32
+// workflow_name          offset  77, size 10
+// workflow_owner         offset  87, size 20
+// report_id              offset 107, size  2
+
+pub fn extract_config_id(raw_report: &[u8]) -> [u8; 8] {
+    // don_id | don_config_version
+    raw_report[37..45].try_into().expect("Expected 8 bytes")
```

Comment:
Another case where this will panic, would it be better to guard this with require!

Reply:
good check, this something we have a ticket for @chray-zhang will pick up some of the small improvements